lib/Transforms/HandshakeOptimization.cpp

```diff
@@ -0,0 +1,629 @@
+        useOwner1 == useOwner2) {
+      handshake::MuxOp muxOp = mlir::cast<handshake::MuxOp>(useOwner1);
+      mlir::Value conditionBr = condBranchOp.getConditionOperand();
+      mlir::Value selectMux = muxOp.getSelectOperand();
+
+      if (conditionBr.getDefiningOp() == selectMux.getDefiningOp() and isa_and_nonnull<handshake::ForkOp>(conditionBr.getDefiningOp())) {
+        mlir::Value dataOperand = condBranchOp.getDataOperand();
+        rewriter.create<handshake::SinkOp>(muxOp->getLoc(), selectMux);
+        rewriter.create<handshake::SinkOp>(condBranchOp->getLoc(), conditionBr);
+        rewriter.replaceOp(muxOp, {dataOperand});
+        rewriter.eraseOp(condBranchOp);
+        return success();
+      }
+    }
+    return failure();
+  }
+};
+
+struct RemoveBranchMergePairs
+    : public OpRewritePattern<handshake::ConditionalBranchOp> {
+  using OpRewritePattern<handshake::ConditionalBranchOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(handshake::ConditionalBranchOp condBranchOp,
+                                PatternRewriter &rewriter) const override {
+    mlir::Operation *useOwner1 = nullptr;
+    mlir::Operation *useOwner2 = nullptr;
+
+    auto it = condBranchOp.getTrueResult().getUsers().begin();
+    if (it != condBranchOp.getTrueResult().getUsers().end()) {
+      useOwner1 = *it;
+    }
+
+    it = condBranchOp.getFalseResult().getUsers().begin();
+    if (it != condBranchOp.getFalseResult().getUsers().end()) {
+      useOwner2 = *it;
+    }
+
+    if (isa_and_nonnull<handshake::MergeOp>(useOwner1) and
+        useOwner1 == useOwner2) {
+      handshake::MergeOp mergeOp = mlir::cast<handshake::MergeOp>(useOwner1);
+      mlir::Value conditionBr = condBranchOp.getConditionOperand();
+      mlir::Value dataOperand = condBranchOp.getDataOperand();
+      rewriter.replaceOp(mergeOp, {dataOperand});
+      rewriter.setInsertionPoint(condBranchOp);
+      rewriter.create<handshake::SinkOp>(condBranchOp->getLoc(), conditionBr);
+      rewriter.eraseOp(condBranchOp);
+      return success();
+    }
+    return failure();
+  }
+};
+
+struct RemoveBranchCMergePairs
+    : public OpRewritePattern<handshake::ConditionalBranchOp> {
+  using OpRewritePattern<handshake::ConditionalBranchOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(handshake::ConditionalBranchOp condBranchOp,
+                                PatternRewriter &rewriter) const override {
+    mlir::Operation *useOwner1 = nullptr;
+    mlir::Operation *useOwner2 = nullptr;
+    auto it = condBranchOp.getTrueResult().getUsers().begin();
+    if (it != condBranchOp.getTrueResult().getUsers().end()) {
+      useOwner1 = *it;
+    }
+
+    it = condBranchOp.getFalseResult().getUsers().begin();
+    if (it != condBranchOp.getFalseResult().getUsers().end()) {
+      useOwner2 = *it;
+    }
+
+    if (isa_and_nonnull<handshake::ControlMergeOp>(useOwner1) and
+        useOwner1 == useOwner2) {
+
+      handshake::ControlMergeOp cMergeOp =
```

Comment:
Same as for the merge, if the control merge has more than 2 operands it cannot be deleted.